// @lblod/ember-rdfa-editor-wikipedia-slug-plugin/addon/services/rdfa-editor-wikipedia-slug-plugin.js
import Service from '@ember/service';
import EmberObject from '@ember/object';
import { task } from 'ember-concurrency';

const EDITOR_CARD_NAME = 'editor-plugins/wikipedia-slug-card';

/**
 * Service responsible for correct annotation of dates
 *
 * @module editor-wikipedia-slug-plugin
 * @class RdfaEditorRelatedUrlPlugin
 * @constructor
 * @extends EmberService
 */

export default class RdfaEditorRelatedUrlPlugin extends Service {

  /**
   * task to handle the incoming events from the editor dispatcher
   *
   * @method execute
   *
   * @param {string} hrId Unique identifier of the event in the hintsRegistry
   * @param {Array} contexts RDFa contexts of the text snippets the event applies on
   * @param {Object} hintsRegistry Registry of hints in the editor
   * @param {Object} editor The RDFa editor instance
   *
   * @public
   */
  execute(hrId, rdfaBlocks, hintsRegistry, editor) {
    hintsRegistry.removeHintsInRdfaBlocks( rdfaBlocks, hrId, "wikipedia-slug-scope");
  
    for( const rdfaBlock of rdfaBlocks ){
      let idx = rdfaBlock.text.toLowerCase().indexOf('hello');
      if( idx !== -1 ) {
        // the hintsregistry needs to know the location with respect to the document
        const absoluteLocation = this.normalizeLocation( [idx, idx + 'hello'.length], rdfaBlock.region );
  
        hintsRegistry.addHint( hrId, "wikipedia-slug-scope", {
          // info for the hintsRegistry
          location: absoluteLocation,
          card: "editor-plugins/wikipedia-slug-card",
          // any content you need to render the component and handle its actions
          info: {
            hrId, hintsRegistry, editor,
            location: absoluteLocation,
          }
        });
      }
    }
  }

  /**
   * Maps location of substring back within reference location
   *
   * @method normalizeLocation
   *
   * @param {[int,int]} [start, end] Location withing string
   * @param {[int,int]} [start, end] reference location
   *
   * @return {[int,int]} [start, end] absolute location
   *
   * @private
   */
  normalizeLocation(location, reference) {
    return [location[0] + reference[0], location[1] + reference[0]];
  }

}
